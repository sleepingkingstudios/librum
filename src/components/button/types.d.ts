import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ButtonShims,
} from './shims';

import type { ColorTypes } from '../types';

type ButtonContents = {
  children: React.ReactNode,
} | {
  icon?: IconDefinition,
  label: string,
};

type ButtonOptions = {
  className?: string,
  children?: React.ReactNode,
  disabled?: boolean,
  icon?: IconDefinition,
  label?: string,
  outline?: boolean,
  size?: ButtonSize,
  type?: ColorTypes,
} & ButtonContents;

type ButtonSize = ("sm" | "md" | "lg");

type ButtonType = ("button" | "reset" | "submit");

export type LinkButtonProps = {
  url: string,
} & ButtonOptions;

export type PlainButtonProps = {
  htmlType?: ButtonType,
  onClick?: () => void,
} & ButtonOptions;

export type ButtonProps = LinkButtonProps | PlainButtonProps;
