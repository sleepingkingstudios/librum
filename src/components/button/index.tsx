import * as React from 'react';

import type { IconDefinition } from '@fortawesome/fontawesome-common-types';
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome';
import { Link } from 'react-router-dom';

import { joinClassNames } from '@utils/react-utils';

import type {
  ButtonProps,
  LinkButtonProps,
  PlainButtonProps,
} from './types';

export type { ButtonProps } from './types';

const generateClassName = ({
  className,
  outline = false,
  size,
  type,
}: {
  className?: string,
  outline: boolean,
  size?: string,
  type?: string,
}): string => joinClassNames(
  'btn',
  outline ? 'btn-outline' : null,
  size? `btn-${size}` : null,
  type ? `btn-${type}` : null,
  className,
);

const renderContents = ({
  children,
  icon,
  label,
}: {
  children?: React.ReactNode,
  icon?: IconDefinition,
  label?: string,
}): React.ReactNode => {
  if (children) { return children; }

  if (label) {
    if (!icon) { return label; }

    return (
      <>
        <Icon className="pr-2" icon={icon} />

        { label }
      </>
    )
  }

  return null;
};

const LinkButton = ({
  children,
  className,
  label,
  icon,
  outline,
  size,
  type,
  url,
}: LinkButtonProps): JSX.Element => {
  const joinedClassName = generateClassName({
    className,
    outline,
    size,
    type,
  });

  return (
    <Link className={joinedClassName} to={url}>
      { renderContents({ children, icon, label }) }
    </Link>
  );
};

const PlainButton = ({
  children,
  className,
  disabled = false,
  htmlType = 'button',
  label,
  icon,
  onClick,
  outline,
  size,
  type,
}: PlainButtonProps): JSX.Element => {
  const joinedClassName = generateClassName({
    className,
    outline,
    size,
    type,
  });

  return (
    <button
      className={joinedClassName}
      disabled={disabled}
      onClick={onClick}
      type={htmlType}
    >
      { renderContents({ children, icon, label }) }
    </button>
  );
};

export const Button = (props: ButtonProps): JSX.Element => {
  const { disabled } = props;

  if (('url' in props) && !disabled) { return (<LinkButton {...props} />); }

  return (<PlainButton {...props} />);
};
