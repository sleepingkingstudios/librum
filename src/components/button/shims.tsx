import * as React from 'react';

export const ButtonShims = () => {
  <>
    <button className="btn btn-outline" />
    <button className="btn btn-primary.btn-outline" />
    <button className="btn btn-danger.btn-outline" />
    <button className="btn btn-info.btn-outline" />
    <button className="btn btn-muted.btn-outline" />
    <button className="btn btn-success.btn-outline" />
    <button className="btn btn-warning.btn-outline" />
    <button className="btn btn-sm" />
    <button className="btn btn-md" />
    <button className="btn btn-lg" />
  </>
};
