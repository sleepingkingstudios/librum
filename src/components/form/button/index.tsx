import * as React from 'react';

import { joinClassNames } from '@utils/react-utils';
import type { UpToTwelveColumns } from '@components/types';

interface IFormButtonProps {
  className?: string;
  children: React.ReactNode;
  disabled?: boolean;
  cols?: UpToTwelveColumns;
  type?: "button" | "reset" | "submit";
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

export const FormButton = ({
  className,
  children,
  disabled = false,
  cols = null,
  type = 'button',
}: IFormButtonProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    'form-button',
    columnSpanClassName({ cols }),
    className,
  );

  return (
    <button
      disabled={disabled}
      className={joinedClassName}
      type={type}
    >
      { children }
    </button>
  );
};
