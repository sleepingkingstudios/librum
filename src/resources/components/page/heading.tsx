import * as React from 'react';
import { isPlainObject } from 'lodash';

import type { ButtonProps as Button } from '@components/button';
import { Heading } from '@components/heading';
import { singularize } from '@utils/inflector';
import { titleCase } from '@utils/text';

type ResourcePageLabelProps = {
  action: string,
  data: Record<string, unknown>,
  label?: string,
  member: boolean,
  resourceName: string,
  singularName?: string,
  status: string,
};

type ResourcePageHeadingProps = {
  buttons?: Button[],
} & ResourcePageLabelProps;

const labelFor = ({
  action,
  data,
  label,
  member,
  resourceName,
  singularName,
  status,
}: ResourcePageLabelProps): string => {
  if (typeof label === 'string' && label.length > 0) { return label; }

  if (!member) {
    return `${titleCase(action)} ${titleCase(resourceName)}`.trim();
  }

  const actionName = status === 'success' ? titleCase(action) : 'Loading';
  const memberName = memberNameFor({ data, resourceName, singularName })

  return `${actionName} ${memberName}`.trim();
};

const memberNameFor = ({
  data,
  resourceName,
  singularName,
}: {
  data: Record<string, unknown>,
  resourceName: string,
  singularName?: string,
}): string => {
  const resourceKey = singularName || singularize(resourceName);

  if (resourceKey in data && isPlainObject(data[resourceKey])) {
    const resource = data[resourceKey] as Record<string, unknown>;

    if ('name' in resource) { return resource.name.toString(); }
  }

  return titleCase(resourceKey);
};

export const ResourcePageHeading = ({
  buttons,
  ...labelOptions
}: ResourcePageHeadingProps): JSX.Element => {
  const label: string = labelFor(labelOptions);

  return (
    <Heading buttons={buttons} size={1}>
      { label }
    </Heading>
  );
};
