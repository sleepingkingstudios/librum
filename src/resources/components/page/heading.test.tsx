import * as React from 'react';

import '@testing-library/jest-dom';

import { ResourcePageHeading } from './heading';
import { ButtonProps as Button } from '@components/button';
import { render } from '@test-helpers/rendering';

describe('<ResourcePageHeading />', () => {
  const resourceName = 'rareBooks';

  describe('with a collection action', () => {
    const action = 'published';
    const data = {};
    const member = false;
    const status = 'uninitialized';
    const defaultOptions = {
      action,
      data,
      member,
      resourceName,
      status,
    };

    it('should render the heading', () => {
      const expected = 'Published Rare Books';
      const { getByRole } = render(<ResourcePageHeading {...defaultOptions} />);
      const heading = getByRole('heading');

      expect(heading).toHaveTextContent(expected);
    });

    it('should not render any buttons', () => {
      const { queryAllByRole } =
        render(<ResourcePageHeading {...defaultOptions} />);
      const rendered = queryAllByRole('button');

      expect(rendered).toHaveLength(0);
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(<ResourcePageHeading {...defaultOptions} />);

      expect(asFragment()).toMatchSnapshot();
    });

    describe('with action: empty string', () => {
      const action = '';
      const options = {
        ...defaultOptions,
        action,
      };

      it('should render the heading', () => {
        const expected = 'Rare Books';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('with buttons: value', () => {
      const buttons: Button[] = [
        {
          label: 'Rock',
          type: 'primary',
        },
        {
          label: 'Paper',
          type: 'success',
        },
        {
          label: 'Scissors',
          type: 'danger',
        },
      ];
      const options = {
        ...defaultOptions,
        buttons,
      };

      it('should render the heading', () => {
        const expected = 'Published Rare Books';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should display the buttons', () => {
        const { getAllByRole } = render(<ResourcePageHeading {...options} />);
        const expected = ['Rock', 'Paper', 'Scissors'];

        const rendered = getAllByRole('button');

        expect(rendered).toHaveLength(3);
        expect(rendered.map(button => button.textContent)).toEqual(expected);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('with label: value', () => {
      const label = 'Custom Label';
      const options = {
        ...defaultOptions,
        label,
      };

      it('should render the heading', () => {
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(label);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with a member action', () => {
    const action = 'publish';
    const data = {};
    const member = true;
    const defaultOptions = {
      action,
      data,
      member,
      resourceName,
    };

    describe('when the status is "uninitialized"', () => {
      const status = 'uninitialized';
      const options = {
        ...defaultOptions,
        status,
      };

      it('should render the heading', () => {
        const expected = 'Loading Rare Book';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should not render any buttons', () => {
        const { queryAllByRole } =
          render(<ResourcePageHeading {...options} />);
        const rendered = queryAllByRole('button');

        expect(rendered).toHaveLength(0);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the status is "loading"', () => {
      const status = 'loading';
      const options = {
        ...defaultOptions,
        status,
      };

      it('should render the heading', () => {
        const expected = 'Loading Rare Book';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should not render any buttons', () => {
        const { queryAllByRole } =
          render(<ResourcePageHeading {...options} />);
        const rendered = queryAllByRole('button');

        expect(rendered).toHaveLength(0);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the status is "failure"', () => {
      const status = 'failure';
      const options = {
        ...defaultOptions,
        status,
      };

      it('should render the heading', () => {
        const expected = 'Loading Rare Book';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should not render any buttons', () => {
        const { queryAllByRole } =
          render(<ResourcePageHeading {...options} />);
        const rendered = queryAllByRole('button');

        expect(rendered).toHaveLength(0);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });
    });

    describe('when the status is "success"', () => {
      const status = 'success';
      const options = {
        ...defaultOptions,
        status,
      };

      it('should render the heading', () => {
        const expected = 'Publish Rare Book';
        const { getByRole } = render(<ResourcePageHeading {...options} />);
        const heading = getByRole('heading');

        expect(heading).toHaveTextContent(expected);
      });

      it('should not render any buttons', () => {
        const { queryAllByRole } =
          render(<ResourcePageHeading {...options} />);
        const rendered = queryAllByRole('button');

        expect(rendered).toHaveLength(0);
      });

      it('should match the snapshot', () => {
        const { asFragment } =
          render(<ResourcePageHeading {...options} />);

        expect(asFragment()).toMatchSnapshot();
      });

      describe('when the data includes a named resource', () => {
        const data = {
          rareBook: {
            name: 'Romance of the Three Kingdoms',
          },
        };
        const options = {
          ...defaultOptions,
          data,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Publish Romance of the Three Kingdoms';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });
    });

    describe('with action: empty string', () => {
      const action = '';

      describe('when the status is "success"', () => {
        const status = 'success';
        const options = {
          ...defaultOptions,
          action,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Rare Book';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });

        describe('when the data includes a named resource', () => {
          const data = {
            rareBook: {
              name: 'Romance of the Three Kingdoms',
            },
          };
          const options = {
            ...defaultOptions,
            action,
            data,
            status,
          };

          it('should render the heading', () => {
            const expected = 'Romance of the Three Kingdoms';
            const { getByRole } = render(<ResourcePageHeading {...options} />);
            const heading = getByRole('heading');

            expect(heading).toHaveTextContent(expected);
          });

          it('should match the snapshot', () => {
            const { asFragment } =
              render(<ResourcePageHeading {...options} />);

            expect(asFragment()).toMatchSnapshot();
          });
        });
      });
    });

    describe('with buttons: value', () => {
      const buttons: Button[] = [
        {
          label: 'Rock',
          type: 'primary',
        },
        {
          label: 'Paper',
          type: 'success',
        },
        {
          label: 'Scissors',
          type: 'danger',
        },
      ];

      describe('when the status is "uninitialized"', () => {
        const status = 'uninitialized';
        const options = {
          ...defaultOptions,
          buttons,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Book';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should display the buttons', () => {
          const { getAllByRole } = render(<ResourcePageHeading {...options} />);
          const expected = ['Rock', 'Paper', 'Scissors'];

          const rendered = getAllByRole('button');

          expect(rendered).toHaveLength(3);
          expect(rendered.map(button => button.textContent)).toEqual(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "loading"', () => {
        const status = 'loading';
        const options = {
          ...defaultOptions,
          buttons,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Book';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should display the buttons', () => {
          const { getAllByRole } = render(<ResourcePageHeading {...options} />);
          const expected = ['Rock', 'Paper', 'Scissors'];

          const rendered = getAllByRole('button');

          expect(rendered).toHaveLength(3);
          expect(rendered.map(button => button.textContent)).toEqual(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "failure"', () => {
        const status = 'failure';
        const options = {
          ...defaultOptions,
          buttons,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Book';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should display the buttons', () => {
          const { getAllByRole } = render(<ResourcePageHeading {...options} />);
          const expected = ['Rock', 'Paper', 'Scissors'];

          const rendered = getAllByRole('button');

          expect(rendered).toHaveLength(3);
          expect(rendered.map(button => button.textContent)).toEqual(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "success"', () => {
        const status = 'success';
        const options = {
          ...defaultOptions,
          buttons,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Publish Rare Book';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should display the buttons', () => {
          const { getAllByRole } = render(<ResourcePageHeading {...options} />);
          const expected = ['Rock', 'Paper', 'Scissors'];

          const rendered = getAllByRole('button');

          expect(rendered).toHaveLength(3);
          expect(rendered.map(button => button.textContent)).toEqual(expected);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });

        describe('when the data includes a named resource', () => {
          const data = {
            rareBook: {
              name: 'Romance of the Three Kingdoms',
            },
          };
          const options = {
            ...defaultOptions,
            buttons,
            data,
            status,
          };

          it('should render the heading', () => {
            const expected = 'Publish Romance of the Three Kingdoms';
            const { getByRole } = render(<ResourcePageHeading {...options} />);
            const heading = getByRole('heading');

            expect(heading).toHaveTextContent(expected);
          });

          it('should match the snapshot', () => {
            const { asFragment } =
              render(<ResourcePageHeading {...options} />);

            expect(asFragment()).toMatchSnapshot();
          });
        });
      });
    });

    describe('with label: value', () => {
      const label = 'Custom Label';

      describe('when the status is "uninitialized"', () => {
        const status = 'uninitialized';
        const options = {
          ...defaultOptions,
          label,
          status,
        };

        it('should render the heading', () => {
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(label);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "loading"', () => {
        const status = 'loading';
        const options = {
          ...defaultOptions,
          label,
          status,
        };

        it('should render the heading', () => {
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(label);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "failure"', () => {
        const status = 'failure';
        const options = {
          ...defaultOptions,
          label,
          status,
        };

        it('should render the heading', () => {
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(label);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "success"', () => {
        const status = 'success';
        const options = {
          ...defaultOptions,
          label,
          status,
        };

        it('should render the heading', () => {
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(label);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });

        describe('when the data includes a named resource', () => {
          const data = {
            rareBook: {
              name: 'Romance of the Three Kingdoms',
            },
          };
          const options = {
            ...defaultOptions,
            data,
            label,
            status,
          };

          it('should render the heading', () => {
            const { getByRole } = render(<ResourcePageHeading {...options} />);
            const heading = getByRole('heading');

            expect(heading).toHaveTextContent(label);
          });

          it('should match the snapshot', () => {
            const { asFragment } =
              render(<ResourcePageHeading {...options} />);

            expect(asFragment()).toMatchSnapshot();
          });
        });
      });
    });

    describe('with singularName: value', () => {
      const singularName = 'rareTome';

      describe('when the status is "uninitialized"', () => {
        const status = 'uninitialized';
        const options = {
          ...defaultOptions,
          singularName,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Tome';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "loading"', () => {
        const status = 'loading';
        const options = {
          ...defaultOptions,
          singularName,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Tome';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "failure"', () => {
        const status = 'failure';
        const options = {
          ...defaultOptions,
          singularName,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Loading Rare Tome';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });
      });

      describe('when the status is "success"', () => {
        const status = 'success';
        const options = {
          ...defaultOptions,
          singularName,
          status,
        };

        it('should render the heading', () => {
          const expected = 'Publish Rare Tome';
          const { getByRole } = render(<ResourcePageHeading {...options} />);
          const heading = getByRole('heading');

          expect(heading).toHaveTextContent(expected);
        });

        it('should not render any buttons', () => {
          const { queryAllByRole } =
            render(<ResourcePageHeading {...options} />);
          const rendered = queryAllByRole('button');

          expect(rendered).toHaveLength(0);
        });

        it('should match the snapshot', () => {
          const { asFragment } =
            render(<ResourcePageHeading {...options} />);

          expect(asFragment()).toMatchSnapshot();
        });

        describe('when the data includes a named resource', () => {
          const data = {
            rareTome: {
              name: 'Romance of the Three Kingdoms',
            },
          };
          const options = {
            ...defaultOptions,
            data,
            singularName,
            status,
          };

          it('should render the heading', () => {
            const expected = 'Publish Romance of the Three Kingdoms';
            const { getByRole } = render(<ResourcePageHeading {...options} />);
            const heading = getByRole('heading');

            expect(heading).toHaveTextContent(expected);
          });

          it('should match the snapshot', () => {
            const { asFragment } =
              render(<ResourcePageHeading {...options} />);

            expect(asFragment()).toMatchSnapshot();
          });
        });
      });
    });
  });
});
