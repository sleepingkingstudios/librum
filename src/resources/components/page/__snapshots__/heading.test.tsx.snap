// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourcePageHeading /> with a collection action should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Published Rare Books
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a collection action with action: empty string should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Rare Books
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a collection action with buttons: value should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Published Rare Books
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a collection action with label: value should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action when the status is "failure" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Book
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action when the status is "loading" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Book
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action when the status is "success" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Publish Rare Book
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action when the status is "success" when the data includes a named resource should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Publish Romance of the Three Kingdoms
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action when the status is "uninitialized" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Book
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with action: empty string when the status is "success" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Rare Book
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with action: empty string when the status is "success" when the data includes a named resource should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Romance of the Three Kingdoms
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with buttons: value when the status is "failure" should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Loading Rare Book
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with buttons: value when the status is "loading" should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Loading Rare Book
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with buttons: value when the status is "success" should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Publish Rare Book
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with buttons: value when the status is "success" when the data includes a named resource should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Publish Romance of the Three Kingdoms
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with buttons: value when the status is "uninitialized" should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col sm:flex-row"
  >
    <h1>
      Loading Rare Book
    </h1>
    <ul
      class="justify-end flex flex-col grow sm:flex-row btn-list"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Rock
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Paper
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Scissors
      </button>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with label: value when the status is "failure" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with label: value when the status is "loading" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with label: value when the status is "success" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with label: value when the status is "success" when the data includes a named resource should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with label: value when the status is "uninitialized" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Custom Label
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with singularName: value when the status is "failure" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Tome
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with singularName: value when the status is "loading" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Tome
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with singularName: value when the status is "success" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Publish Rare Tome
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with singularName: value when the status is "success" when the data includes a named resource should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Publish Romance of the Three Kingdoms
  </h1>
</DocumentFragment>
`;

exports[`<ResourcePageHeading /> with a member action with singularName: value when the status is "uninitialized" should match the snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading Rare Tome
  </h1>
</DocumentFragment>
`;
